# Dataset: AB_NYC_2019

### **Author: Aregbesola Emmanuel**
### **Date: March 2026**

=====================================================================================================

In [1]:
# import pandas library

import pandas as pd

In [2]:
# load the data into pandas

df = pd.read_csv('AB_NYC_2019.csv')
pd.set_option('display.max_rows', 100)

# 1. Loading and Initial Inspection

In this section, we are going to do basic inspection on our dataset. This is to get a feel of what our dataset looks like and how we are going to approach the cleaning.

We are going to check for:
1. Shape.
2. Column Names.
3. Data Types (info()).
4. Summary Statistics (describe()).
5. Duplicated Rows.
6. Null values.
7. Null Percentage Per Column.

=====================================================================================================

In [3]:
# check the number of rows and columns
df.shape

(48895, 16)

In [4]:
# check the column names for correct formatting

df.columns.to_list()

['id',
 'name',
 'host_id',
 'host_name',
 'neighbourhood_group',
 'neighbourhood',
 'latitude',
 'longitude',
 'room_type',
 'price',
 'minimum_nights',
 'number_of_reviews',
 'last_review',
 'reviews_per_month',
 'calculated_host_listings_count',
 'availability_365']

In [5]:
# check the data type of each columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [6]:
# summary statistics across all columns
df.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


In [7]:
# check for duplicated rows
df.duplicated().sum()

0

In [8]:
# check for null values across all columns
df.isna().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [9]:
# percentage of null values per column

for col in df.columns:
    col_null = df[col].isna().sum()
    print(f'{col} has {(col_null/len(df))*100:.3f}% null values')

id has 0.000% null values
name has 0.033% null values
host_id has 0.000% null values
host_name has 0.043% null values
neighbourhood_group has 0.000% null values
neighbourhood has 0.000% null values
latitude has 0.000% null values
longitude has 0.000% null values
room_type has 0.000% null values
price has 0.000% null values
minimum_nights has 0.000% null values
number_of_reviews has 0.000% null values
last_review has 20.558% null values
reviews_per_month has 20.558% null values
calculated_host_listings_count has 0.000% null values
availability_365 has 0.000% null values


# Summary

After the initial inspection, these are our findings;

1. We have 48k rows and 16 columns
2. All columns in the dataset are correctly formatted
3. There are no duplicate rows
4. We have null values in name, host_name, last_review, and review_per_month columns

=====================================================================================================

# 2. Column by Column Exploration

In this section, we will go through all the columns in this dataset **individually**. 

=====================================================================================================

## 2.1. Id Column

### Expectation?
The id column is the unique identifer for listing. 
We expect that the column should have distinct value -- No duplicates.

Also, we expect no null values

In [10]:
# Inspect the Id column

print(f'id has {df["id"].isna().sum()} null values')
print(f'id has {df["id"].duplicated().sum()} duplicate rows')
print(f'id: {df["id"].dtype}')

id has 0 null values
id has 0 duplicate rows
id: int64


## Id Column Cleaning

No Null Values. No Duplicates. Correct Data Type.

Note: If duplicates had been found in the column, it would be investigated before dropping. duplicate Id could be data entry error amongst others.

=====================================================================================================

## 2.2. Name Column

### Expectation?

The name column is the listing name for each listing. It should be a text based column.

In section 1, we saw that there is a 0.033% of null values in the columnns.

In [11]:
# Inspect the name column

print(f'name has {df["name"].isna().sum()} null values')
print(f'name has {df["name"].duplicated().sum()} duplicate rows')
print(f'name: {df["name"].dtype}')

name has 16 null values
name has 989 duplicate rows
name: object


In [12]:
# filter out the null values as a dataframe

name_null_mask = df['name'].isna()
df.loc[name_null_mask]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
2854,1615764,NaN,6676776,Peter,Manhattan,Battery Park City,40.71239,-74.01620,Entire home/apt,400,1000,0,NaN,NaN,1,362
3703,2232600,NaN,11395220,Anna,Manhattan,East Village,40.73215,-73.98821,Entire home/apt,200,1,28,2015-06-08,0.45,1,341
5775,4209595,NaN,20700823,Jesse,Manhattan,Greenwich Village,40.73473,-73.99244,Entire home/apt,225,1,1,2015-01-01,0.02,1,0
5975,4370230,NaN,22686810,Michaël,Manhattan,Nolita,40.72046,-73.99550,Entire home/apt,215,7,5,2016-01-02,0.09,1,0
6269,4581788,NaN,21600904,Lucie,Brooklyn,Williamsburg,40.71370,-73.94378,Private room,150,1,0,NaN,NaN,1,0
6567,4756856,NaN,1832442,Carolina,Brooklyn,Bushwick,40.70046,-73.92825,Private room,70,1,0,NaN,NaN,1,0
6605,4774658,NaN,24625694,Josh,Manhattan,Washington Heights,40.85198,-73.93108,Private room,40,1,0,NaN,NaN,1,0
8841,6782407,NaN,31147528,Huei-Yin,Brooklyn,Williamsburg,40.71354,-73.93882,Private room,45,1,0,NaN,NaN,1,0
11963,9325951,NaN,33377685,Jonathan,Manhattan,Hell's Kitchen,40.76436,-73.98573,Entire home/apt,190,4,1,2016-01-05,0.02,1,0
12824,9787590,NaN,50448556,Miguel,Manhattan,Harlem,40.80316,-73.95189,Entire home/apt,300,5,0,NaN,NaN,5,0


In [13]:
# check for stale listing

df.loc[name_null_mask, ['name', 'number_of_reviews', 'availability_365', 'price']].sort_values('number_of_reviews')

,name,number_of_reviews,availability_365,price
2854,NaN,0,362,400
6269,NaN,0,0,150
6567,NaN,0,0,70
6605,NaN,0,0,40
8841,NaN,0,0,45
12824,NaN,0,0,300
13059,NaN,0,0,67
15819,NaN,0,0,100
16071,NaN,0,365,130
18047,NaN,0,0,70


In [14]:
# filter out the stale listings

stale_mask = ((df['number_of_reviews'] == 0) 
              & (df['availability_365'] == 0)
             & df['name'].isna())
    
df[stale_mask].shape

(8, 16)

In [15]:
# drop the stale listings

df = df[~stale_mask].reset_index(drop=True)
df.shape

(48887, 16)

In [16]:
# fill the null values in name column with 'unknown name'

df['name'] = df['name'].fillna('Unknown')

In [17]:
# Validate the fill

df['name'].isna().sum()

0

In [18]:
# filter out the duplicates in the name column

dupes_name_mask = ((df['number_of_reviews'] == 0) &
                   (df.duplicated(subset=['name', 'host_id', 'neighbourhood', 
                                        'room_type'], keep=False))
)



In [19]:
# view filter

df.loc[dupes_name_mask, ['name', 'host_id', 'neighbourhood', 
                                        'room_type', 'number_of_reviews']].sort_values('name')

,name,host_id,neighbourhood,room_type,number_of_reviews
42768,1 BR Suite in Midtown Manhattan,24831061,Midtown,Private room,0
42809,1 BR Suite in Midtown Manhattan,24831061,Midtown,Private room,0
30694,1 bedroom apartment on Columbus Avenue,76104209,Upper West Side,Entire home/apt,0
31992,1 bedroom apartment on Columbus Avenue,76104209,Upper West Side,Entire home/apt,0
33159,20-25 minutes to downtown NYC/30-40 to midtown,89769986,Bedford-Stuyvesant,Private room,0
...,...,...,...,...,...
22749,"West 74th Street, Svcd Studio Apartment",22541573,Upper West Side,Entire home/apt,0
37721,Your Manhattan Home: 12 mins to NYC,213781715,Greenpoint,Private room,0
47957,★Premier Queen Room with Balcony ★,270874051,Long Island City,Private room,0
48018,★Premier Queen Room with Balcony ★,270874051,Long Island City,Private room,0


In [20]:
# drop filter

df.sort_values('id', inplace=True)

df = df[
    ~dupes_name_mask |
    ~df.duplicated(subset=['name', 'host_id', 'neighbourhood', 'room_type'], keep='first')
].reset_index(drop=True)

In [21]:
print(f'{df.columns[1]} has {df[df.columns[1]].isna().sum()} null values')
print(f'{df.columns[1]} has {df[df.columns[1]].duplicated().sum()} duplicate rows')
print(f'{df.columns[1]}: {df[df.columns[1]].dtype}')
df.shape

name has 0 null values
name has 877 duplicate rows
name: object


(48783, 16)

## Name Column Cleaning

### Null Values
The `name` column had 16 missing values. Upon inspection, 8 of these listings had zero reviews and zero availability — indicating stale/abandoned listings with no transaction history and no active availability on the platform. These 8 rows were dropped.

The remaining 8 null listings had either reviews or availability, suggesting they were real active listings. Since the `name` column is descriptive and not used in analysis, nulls were filled with `'Unknown'` to preserve these rows.

### Duplicates
Duplicates were defined as rows sharing the same `name`, `host_id`, `neighbourhood`, and `room_type` — with `number_of_reviews == 0` on all occurrences. Listings with reviews were excluded from this definition as they represent real transaction history 
and are considered legitimate separate listings on the platform.    
877 remaining duplicate names are expected — these are legitimate listings from different or same hosts with organically similar names and confirmed review history.

The `id` column was used as a tiebreaker — the lowest `id` (original listing) was kept and all subsequent duplicates were dropped.

**Duplicates before cleaning:** 989     
**Duplicates after cleaning:** 877     
**Duplicates dropped:** 104

**Rows before cleaning:** 48,895    
**Rows after cleaning:** 48,783  
**Rows dropped:** 112 (8 stale listings and 104 duplicates)

=====================================================================================================

## 2.3. Host_Id Column

### Expectation?
The `host_id` is a unique identifier given to each host on the platform. We can expect to have duplicate `host_id` as oppose to our `id` since a single host can decide to have multiple listings.      
From our basic inspection, we have seen that there are no null values and the data type is an `object`.

In [22]:
# Inspect the host_id column

print(f'host_id has {df["host_id"].isna().sum()} null values')
print(f'host_id has {df["host_id"].duplicated().sum()} duplicate rows')
print(f'host_id: {df["host_id"].dtype}')

host_id has 0 null values
host_id has 11332 duplicate rows
host_id: int64


In [23]:
# mapping host_id to host_name

host_mapping = df.groupby('host_id')['host_name'].nunique().sort_values(ascending=False)
host_mapping[host_mapping > 1]

Series([], Name: host_name, dtype: int64)

## Host_Id Column Cleaning

We expected that there would be duplicate host_id in the dataset because a host can have multiple listings. To confirm if the duplicates are not as a result of error, we have decided to map every `host_id` to a `host_name` to check if every id belongs to one name.
If the mapping had resulted to some `host_id` belonging to more than one name, more investigation would been done to why.

=====================================================================================================

## 2.4. Host_Name Column

### Expectation?
`host_name` is the display name for hosts on the platform. Duplicates are expected and normal for this column since hosts can bear the same name and a single host can have multiple listings.   
In our basic inspection, we can expect to see 21 null values.

In [24]:
# Inspect the host_name column

print(f'host_name has {df["host_name"].isna().sum()} null values')
print(f'host_name has {df["host_name"].duplicated().sum()} duplicate rows')
print(f'host_name: {df["host_name"].dtype}')

host_name has 21 null values
host_name has 37331 duplicate rows
host_name: object


In [25]:
# Filter out the null value from the main dataframe

host_name_mask = df['host_name'].isna()
df.loc[host_name_mask]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
360,100184,Bienvenue,526653,NaN,Queens,Queens Village,40.72413,-73.76133,Private room,50,1,43,2019-07-08,0.45,1,88
2699,1449546,Cozy Studio in Flatbush,7779204,NaN,Brooklyn,Flatbush,40.64965,-73.96154,Entire home/apt,100,30,49,2017-01-02,0.69,1,342
5744,4183989,SPRING in the City!! Zen-Style Tranquil Bedroom,919218,NaN,Manhattan,Harlem,40.80606,-73.95061,Private room,86,3,34,2019-05-23,1.00,1,359
6074,4446862,Charming Room in Prospect Heights!,23077718,NaN,Brooklyn,Crown Heights,40.67512,-73.96146,Private room,50,1,0,NaN,NaN,1,0
6579,4763327,"Luxurious, best location, spa inc'l",24576978,NaN,Brooklyn,Greenpoint,40.72035,-73.95355,Entire home/apt,195,1,1,2015-10-20,0.02,1,0
8159,6292866,Modern Quiet Gem Near All,32722063,NaN,Brooklyn,East Flatbush,40.65263,-73.93215,Entire home/apt,85,2,182,2019-06-19,3.59,2,318
8253,6360224,"Sunny, Private room in Bushwick",33134899,NaN,Brooklyn,Bushwick,40.70146,-73.92792,Private room,37,1,1,2015-07-01,0.02,1,0
8847,6786181,R&S Modern Spacious Hideaway,32722063,NaN,Brooklyn,East Flatbush,40.64345,-73.93643,Entire home/apt,100,2,157,2019-06-19,3.18,2,342
9133,6992973,1 Bedroom in Prime Williamsburg,5162530,NaN,Brooklyn,Williamsburg,40.71838,-73.95630,Entire home/apt,145,1,0,NaN,NaN,1,0
9812,7556587,Sunny Room in Harlem,39608626,NaN,Manhattan,Harlem,40.82929,-73.94182,Private room,28,1,1,2015-08-01,0.02,1,0


In [26]:
# Check for stale listing

host_name_stale = ((df['host_name'].isna()) &
                  (df['number_of_reviews'] == 0) &
                  (df['availability_365'] == 0))
df.loc[host_name_stale]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
6074,4446862,Charming Room in Prospect Heights!,23077718,NaN,Brooklyn,Crown Heights,40.67512,-73.96146,Private room,50,1,0,NaN,NaN,1,0
9133,6992973,1 Bedroom in Prime Williamsburg,5162530,NaN,Brooklyn,Williamsburg,40.71838,-73.95630,Entire home/apt,145,1,0,NaN,NaN,1,0
27729,21787094,Spacious and Bright Midtown East Apartment,415290,NaN,Manhattan,Upper East Side,40.76436,-73.96836,Entire home/apt,325,1,0,NaN,NaN,1,0
32140,25180572,Beautiful Cozy Studio Apt. UWS,119609345,NaN,Manhattan,Upper West Side,40.77701,-73.97667,Entire home/apt,180,4,0,NaN,NaN,1,0


In [27]:
# drop the stale listing

df = df[~host_name_stale].reset_index(drop=True)
df.shape

(48779, 16)

In [28]:
# map host_name to host_id from existing records and fill null host_names

host_name_to_id = df.dropna(subset=['host_name']).groupby('host_id')['host_name'].first()
df['host_name'] = df['host_name'].fillna(df['host_id'].map(host_name_to_id))

In [29]:
# check if we still have missing values

df['host_name'].isna().sum()

17

In [30]:
# Fill the null values in host_name with 'unknown'

df['host_name'] = df['host_name'].fillna('unknown')

In [31]:
# validate the fill

df['host_name'].isna().sum()

0

In [32]:
# Validate the host_name column

print(f'{df.columns[3]} has {df[df.columns[3]].isna().sum()} null values')
print(f'{df.columns[3]} has {df[df.columns[3]].duplicated().sum()} duplicate rows')
print(f'{df.columns[3]}: {df[df.columns[3]].dtype}')
df.shape

host_name has 0 null values
host_name has 37327 duplicate rows
host_name: object


(48779, 16)

## Host_Name Column Cleaning

After inspecting the null values in the `host_name` column, we realized that there were some stale listings. These were listings that do not have a `host_name`, no `number_of_reviews`, and no `availability_365`. We dropped the stale listings.    

We went on to try to map host_name to host_id from the existing records. This helps to check if there is a host_name for the same host_id in our existing record. Unfortunately, there wasn't.   
We decided to fill the null values in the `host_name` column with **unknown**


**Rows before cleaning:** 48,783   
**Rows after cleaning:** 48,779   
**Rows dropped:** 4

=====================================================================================================

## 2.5. Neighbourhood_group Column

### Expectation?
`neighbourhood_group` represents the borough in New York City where the listing is located.    
NYC has exactly 5 boroughs: Manhattan, Brooklyn, Queens, The Bronx, and Staten Island.    
Every listing should belong to one of these 5 values — no nulls, no unexpected categories.   
From our basic inspection, there are no null values, duplicates are expected.

In [33]:
# Inspect the neighbourhood_group column

print(f'neighbourhood_group has {df["neighbourhood_group"].isna().sum()} null values')
print(f'neighbourhood_group has {df["neighbourhood_group"].duplicated().sum()} duplicate rows')
print(f'neighbourhood_group: {df["neighbourhood_group"].dtype}')

neighbourhood_group has 0 null values
neighbourhood_group has 48774 duplicate rows
neighbourhood_group: object


In [34]:
# validate if there are no unexpexted categories in the column

df['neighbourhood_group'].value_counts()

Manhattan        21593
Brooklyn         20076
Queens            5646
Bronx             1091
Staten Island      373
Name: neighbourhood_group, dtype: int64

## Neighbourhood_group Column Cleaning

No null values.   
No unexpected categories.  
The column is clean, no cleaning required.

=====================================================================================================

## 2.6. Neighbourhood Column

### Expectation?
`neighbourhood` represents the specific neighbourhood within each borough where the listing 
is located. Multiple listings can share the same neighbourhood which means duplicates are expected and 
normal.     
From basic inspection, no nulls expected. Values should be consistent with their corresponding `neighbourhood_group` (borough).

In [35]:
# Inspect the neighbourhood column

print(f'neighbourhood has {df["neighbourhood"].isna().sum()} null values')
print(f'neighbourhood has {df["neighbourhood"].duplicated().sum()} duplicate rows')
print(f'neighbourhood: {df["neighbourhood"].dtype}')

neighbourhood has 0 null values
neighbourhood has 48558 duplicate rows
neighbourhood: object


In [36]:
# Check if neighbourhood values corresponds with their borough

df.groupby('neighbourhood')['neighbourhood_group'].nunique().sort_values(ascending=False).head(10)

neighbourhood
Allerton           1
Parkchester        1
New Dorp           1
New Dorp Beach     1
New Springville    1
NoHo               1
Nolita             1
North Riverdale    1
Norwood            1
Oakwood            1
Name: neighbourhood_group, dtype: int64

In [37]:
# check for nunmber of unique neighbourhood

df['neighbourhood'].nunique()

221

## Neighbourhood Column Cleaning

No null values.   
221 unique neighbourhood.   
Each neighbourhood exactly tied to one borough.   
No cleaning needed.

=====================================================================================================

## 2.7. Latitude Column

### Expectation?
`latitude` represents the geographic coordinate of each listing. No nulls expected. 
Duplicates are expected — different listings at different addresses will have unique coordinates, 
but listings in the same building could share coordinates.

NYC latitude range is approximately **40.4774 to 40.9176**. Any value outside this range 
is invalid and should be investigated.

In [38]:
# Inspect latitude column

print(f'latitude has {df["latitude"].isna().sum()} null values')
print(f'latitude has {df["latitude"].duplicated().sum()} duplicate rows')
print(f'latitude: {df["latitude"].dtype}')

latitude has 0 null values
latitude has 29737 duplicate rows
latitude: float64


In [39]:
# Check for latittude range

print(f'{df["latitude"].min()}, {df["latitude"].max()}')

40.49979, 40.91306


## Latitude Column Cleaning

No null values.   
Falls within NYC bounds.      
No cleaning required.

=====================================================================================================

## 2.8. Longitude Column

### Expectation?

`longitude` represents the geographic coordinate of each listing. No nulls expected. 
Duplicates are expected — different listings at different addresses will have unique coordinates, 
but listings in the same building could share coordinates.
NYC longitude range is approximately **-74.2591 to -73.7004**. Any value outside this range 
is invalid and should be investigated.

In [40]:
# Inspect longitude column

print(f'longitude has {df["longitude"].isna().sum()} null values')
print(f'longitude has {df["longitude"].duplicated().sum()} duplicate rows')
print(f'longitude: {df["longitude"].dtype}')

longitude has 0 null values
longitude has 34077 duplicate rows
longitude: float64


In [41]:
# Check for longitude range

print(f'{df["longitude"].min()}, {df["longitude"].max()}')

-74.24442, -73.71299


## Longitude Column Cleaning

No null values.   
Falls within NYC bounds.      
No cleaning required.

=====================================================================================================

## 2.9. Room_type Column

### Expectation?
`room_type` represents the type of room the listing is.    
air_bnb has exactly 3 `room_type`: Entire Home/apt, Private Room, and Shared Room.   
Every listing should belong to one of these 3 `room_type` — no nulls, no unexpected categories.   
From our basic inspection, there are no null values, duplicates are expected.

In [42]:
# Inspect room_type column

print(f'room_type has {df["room_type"].isna().sum()} null values')
print(f'room_type has {df["room_type"].duplicated().sum()} duplicate rows')
print(f'room_type: {df["room_type"].dtype}')

room_type has 0 null values
room_type has 48776 duplicate rows
room_type: object


In [43]:
# check for unexpected categories

df['room_type'].value_counts()

Entire home/apt    25351
Private room       22271
Shared room         1157
Name: room_type, dtype: int64

## Room_type Column Cleaning

There are no null values.    
There are no unexpected categories.
No cleaning required.

=====================================================================================================

## 2.10. Price Column

### Expectation?
`price` represents the price value that the listing costs.   
There should be no negative values in this column.    
Also, there should be no price that is 0, that could be a data error.   
From our basic inspection, we have no null values, the data type is integer.    
We can expect there should be duplicates.

In [44]:
# Inspect price column

print(f'price has {df["price"].isna().sum()} null values')
print(f'price has {df["price"].duplicated().sum()} duplicate rows')
print(f'price: {df["price"].dtype}')

price has 0 null values
price has 48106 duplicate rows
price: int64


In [45]:
# check for negative prices

negative_price = df['price'] < 0
df.loc[negative_price]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365


In [46]:
# check for prices that are zero

zero_price = df['price'] == 0
df.loc[zero_price]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
23138,18750597,"Huge Brooklyn Brownstone Living, Close to it all.",8993084,Kimberly,Brooklyn,Bedford-Stuyvesant,40.69023,-73.95428,Private room,0,4,1,2018-01-06,0.05,4,28
25407,20333471,★Hostel Style Room | Ideal Traveling Buddies★,131697576,Anisha,Bronx,East Morrisania,40.83296,-73.88668,Private room,0,2,55,2019-06-24,2.56,4,127
25607,20523843,"MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)",15787004,Martial Loft,Brooklyn,Bushwick,40.69467,-73.92433,Private room,0,2,16,2019-05-18,0.71,5,0
25726,20608117,"Sunny, Quiet Room in Greenpoint",1641537,Lauren,Brooklyn,Greenpoint,40.72462,-73.94072,Private room,0,2,12,2017-10-27,0.53,2,0
25751,20624541,Modern apartment in the heart of Williamsburg,10132166,Aymeric,Brooklyn,Williamsburg,40.70838,-73.94645,Entire home/apt,0,5,3,2018-01-02,0.15,1,73
25767,20639628,Spacious comfortable master bedroom with nice ...,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68173,-73.91342,Private room,0,1,93,2019-06-15,4.28,6,176
25768,20639792,Contemporary bedroom in brownstone with nice view,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68279,-73.91170,Private room,0,1,95,2019-06-21,4.37,6,232
25769,20639914,Cozy yet spacious private brownstone bedroom,86327101,Adeyemi,Brooklyn,Bedford-Stuyvesant,40.68258,-73.91284,Private room,0,1,95,2019-06-23,4.35,6,222
26223,20933849,the best you can find,13709292,Qiuchi,Manhattan,Murray Hill,40.75091,-73.97597,Entire home/apt,0,3,0,NaN,NaN,1,0
26803,21291569,Coliving in Brooklyn! Modern design / Shared room,101970559,Sergii,Brooklyn,Bushwick,40.69211,-73.90670,Shared room,0,30,2,2019-06-22,0.11,6,333


In [47]:
# drop the zero price

df = df[~zero_price].reset_index(drop=True)


In [48]:
df.shape

(48768, 16)

In [49]:
# validate price column

print(f'{df.columns[9]} has {df[df.columns[9]].isna().sum()} null values')
print(f'{df.columns[9]} has {df[df.columns[9]].duplicated().sum()} duplicate rows')
print(f'{df.columns[9]}: {df[df.columns[9]].dtype}')

price has 0 null values
price has 48096 duplicate rows
price: int64


## Price Column Cleaning

11 listings (0.023% of the dataset) have a price of $0. Since price is a core metric in this dataset, retaining $0 values would incorrectly pull down mean price calculations and distort any price-based segmentation or analysis. The correct price cannot be recovered from available columns, so these rows were dropped.

**Rows before cleaning:** 48,779   
**Rows after cleaning:**  48,768    
**Rows dropped:** 11

=====================================================================================================

## 2.11. Minimum_nights Column

### Expectation?
`minimum_nights` represents the minimum number of nights a customer must book for a listing. 
Duplicates are expected. No nulls expected. Values should be positive integers — zero and 
negative values are invalid. Extremely high values (365+) are suspicious and worth investigating.

In [50]:
# inspect minimum_night column

print(f'minimum_nights has {df["minimum_nights"].isna().sum()} null values')
print(f'minimum_nights has {df["minimum_nights"].duplicated().sum()} duplicate rows')
print(f'minimum_nights: {df["minimum_nights"].dtype}')

minimum_nights has 0 null values
minimum_nights has 48659 duplicate rows
minimum_nights: int64


In [51]:
# check for extreme values

print(f'{df["minimum_nights"].min()}, {df["minimum_nights"].max()}')

1, 1250


In [52]:
# check for negative values

print((df['minimum_nights'] <= 0).sum())

0


## Minimun Nights Column Cleaning  

No null values. No zero or negative values — minimum is 1. Maximum is 1250, which is 
unusually high but cannot be confirmed as a data error without further investigation.    
Flagged for EDA. No cleaning required.

=====================================================================================================

## 2.12. Number_of_reviews Column

### Expectation?
`number_of_reviews` represents the total amount of reviews customer have given a particular listing.    
There should be no null values.   
Duplicates are expected.   
There should be no negative values in the column.   


In [53]:
# inspect number_of_review column

print(f'number_of_reviews has {df["number_of_reviews"].isna().sum()} null values')
print(f'number_of_reviews has {df["number_of_reviews"].duplicated().sum()} duplicate rows')
print(f'number_of_reviews: {df["number_of_reviews"].dtype}')

number_of_reviews has 0 null values
number_of_reviews has 48374 duplicate rows
number_of_reviews: int64


In [54]:
# check min and max number of reviews

print(f" {df['number_of_reviews'].min()}, {df['number_of_reviews'].max()}")

 0, 629


## Number_of_reviews Column Cleaning

No null values. No negative values — minimum is 0 (valid for new listings).    
Maximum is 629. No cleaning required.

=====================================================================================================

## 2.13. Last_review Column

### Expectation?
`last_review` represents the date a review was last left for a listing. The data type should 
be datetime. Null values are expected and valid — listings with zero reviews will have no 
last review date. Any listing with a review date but zero reviews would be a data integrity 
issue worth investigating.

In [55]:
# inspect last_review column

print(f'last_review has {df["last_review"].isna().sum()} null values')
print(f'last_review has {df["last_review"].duplicated().sum()} duplicate rows')
print(f'last_review: {df["last_review"].dtype}')

last_review has 9935 null values
last_review has 47003 duplicate rows
last_review: object


In [56]:
# convert data type to datetime

df['last_review'] = pd.to_datetime(df['last_review'])

In [57]:
# validate the datetime change

df['last_review'].dtype

dtype('<M8[ns]')

In [58]:
# data integrity check

data_check = ( (df['number_of_reviews'] == 0) &
              (df['last_review'].notna())
)
df.loc[data_check]

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365


In [59]:
# validate last_review column

print(f'{df.columns[12]} has {df[df.columns[12]].isna().sum()} null values')
print(f'{df.columns[12]} has {df[df.columns[12]].duplicated().sum()} duplicate rows')
print(f'{df.columns[12]}: {df[df.columns[12]].dtype}')

last_review has 9935 null values
last_review has 47003 duplicate rows
last_review: datetime64[ns]


## Last_review Column Cleaning

Converted `last_review` to datetime. Null values present — all correspond to listings with 
zero reviews, which is expected and valid. No listings with a review date and zero reviews. 
No cleaning required beyond dtype conversion.

=====================================================================================================

## 2.14. Reveiws_per_month Column

### Expectation?
`reviews_per_month` is the average reviews a listing has per month.   
Null values are expected and valid — listings with zero reviews and no last_review will have no 
reviews_per_month.   
Any listing that has reviews_per_month and has both zero reviews and no last review should be investigated.   
Duplicates are expected.

In [60]:
# inspect review_per_month column

print(f'reviews_per_month has {df["reviews_per_month"].isna().sum()} null values')
print(f'reviews_per_month has {df["reviews_per_month"].duplicated().sum()} duplicate rows')
print(f'reviews_per_month: {df["reviews_per_month"].dtype}')

reviews_per_month has 9935 null values
reviews_per_month has 47830 duplicate rows
reviews_per_month: float64


In [61]:
# integrity check

mismatch = df[
    df['reviews_per_month'].notna() & 
    (df['number_of_reviews'] == 0) & 
    (df['last_review'].isna())
]
print(mismatch.shape)

(0, 16)


## Reviews_per_month Column Cleaning


Null values present — all correspond to listings with     
zero reviews, and last_review, which is expected and valid.     
No cleaning required.

=====================================================================================================

## 2.15. Calculated_host_listings_count Column

### Expectation?
`calculated_host_listings_count` represents the number of listings that host has on Airbnb.   
From basic inspection, there are no null values.     
There should be no zero or negative values in this column.    
The same `calculated_host_listings_count` value should be consistent across all rows sharing the same `host_id`

In [62]:
# inspect calculated_host_listings_count column

print(f'calculated_host_listings_count has {df["calculated_host_listings_count"].isna().sum()} null values')
print(f'calculated_host_listings_count has {df["calculated_host_listings_count"].duplicated().sum()} duplicate rows')
print(f'calculated_host_listings_count: {df["calculated_host_listings_count"].dtype}')

calculated_host_listings_count has 0 null values
calculated_host_listings_count has 48721 duplicate rows
calculated_host_listings_count: int64


In [63]:
# validity check 

print(f"{df['calculated_host_listings_count'].min()}, {df['calculated_host_listings_count'].max()}")

1, 327


In [64]:
# integrity check

host_count_check = df.groupby('host_id')['calculated_host_listings_count'].nunique()
print(host_count_check[host_count_check > 1])

Series([], Name: calculated_host_listings_count, dtype: int64)


## Calculated_host_listings_count Column Cleaning

No null values. No zero or negative values — minimum is 1. Maximum is 327, representing 
a professional host with many listings.    
Every host_id maps to exactly one count value —
no integrity issues. No cleaning required.

=====================================================================================================

## 2.16. Availability_365 Column

### Expectation?
`availability_365` represents how many days a listing is available in a year.     
There should be no negative values. Also the maximum day should be 365 else that row should be investigated.    
Duplicates are expected.

In [65]:
# inspect availability_365 column

print(f'availability_365 has {df["availability_365"].isna().sum()} null values')
print(f'availability_365 has {df["availability_365"].duplicated().sum()} duplicate rows')
print(f'availability_365: {df["availability_365"].dtype}')

availability_365 has 0 null values
availability_365 has 48402 duplicate rows
availability_365: int64


In [66]:
# validity check

print(f" {df['availability_365'].min()}, {df['availability_365'].max()}")

 0, 365


## Availability_365 Column Cleaning

No null values. No negative values. Maximum is within the valid 365-day range. 
No cleaning required.

# 3. Database Validation Check

In [67]:
# validate the datatype of the cleaned database

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48768 entries, 0 to 48767
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48768 non-null  int64         
 1   name                            48768 non-null  object        
 2   host_id                         48768 non-null  int64         
 3   host_name                       48768 non-null  object        
 4   neighbourhood_group             48768 non-null  object        
 5   neighbourhood                   48768 non-null  object        
 6   latitude                        48768 non-null  float64       
 7   longitude                       48768 non-null  float64       
 8   room_type                       48768 non-null  object        
 9   price                           48768 non-null  int64         
 10  minimum_nights                  48768 non-null  int64         
 11  nu

In [68]:
# check the shape of the new database

df.shape

(48768, 16)

In [69]:
# check the summary statistics

df.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.876800e+04,4.876800e+04,48768.000000,48768.000000,48768.000000,48768.000000,48768.000000,38833.000000,48768.000000,48768.000000
mean,1.900155e+07,6.750893e+07,40.728939,-73.952182,152.647556,7.006049,23.327346,1.373128,7.021448,112.485277
std,1.098377e+07,7.854996e+07,0.054565,0.046137,239.966651,20.519770,44.589808,1.680391,32.530641,131.455071
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,10.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.458954e+06,7.798076e+06,40.690027,-73.983072,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.965461e+07,3.072157e+07,40.723050,-73.955660,105.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.913342e+07,1.074344e+08,40.763130,-73.936287,175.000000,5.000000,24.000000,2.020000,2.000000,226.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


## Validation

The cleaned dataset contains 48,768 rows and 16 columns — down from 48,895 after removing 127 rows across three cleaning operations.      

All data types are confirmed correct: `last_review` has been converted from object to datetime64[ns], numeric columns retain their int64 and float64 types, and categorical columns remain as object.      

Null values are present only in `last_review` (9,935) and `reviews_per_month` (9,935) — both are expected and valid, corresponding exclusively to listings with zero reviews.         
All other 14 columns are fully populated.


Two items are flagged for investigation in EDA: `minimum_nights` has a maximum of 1,250 — unusually high and likely a data entry error — and `price` has a maximum of $10,000 which may represent outliers that could skew price-based analysis.

## Cleaning Summary

| | Count |
|---|---|
| **Original rows** | 48,895 |
| **Rows dropped** | 127 |
| **Final rows** | 48,768 |
| **Columns** | 16 |

### Changes Made
- Dropped 8 stale listings from `name` column (null name, zero reviews, zero availability)
- Dropped 104 duplicate listings from `name` column (same name, host, neighbourhood, room type, zero reviews)
- Dropped 4 stale listings from `host_name` column
- Filled remaining null `host_name` values via host_id mapping, remainder filled with 'Unknown'
- Filled null `name` values with 'Unknown'
- Converted `last_review` to datetime
- Dropped 11 zero-price listings


### NOTE
The basic EDA will be done in the next notebook -- `AB_NYC_2019 EDA`.                                      
In that notebook, we will look at the price distributions and other explorative analysis.